rust/tw_evm/src/abi/non_empty_array.rs

```diff
@@ -56,6 +56,27 @@ impl<T> NonEmptyArray<T> {
     pub fn into_vec(self) -> Vec<T> {
         self.0
     }
+
+    pub fn get(&self) -> &Vec<T> {
+        &self.0
+    }
+
+    pub fn set(&mut self, value: Vec<T>) {
+        if value.is_empty() {
+            panic!("`NonEmptyArray` must have at least one element");
+        }
+        self.0 = value;
+    }
+
+    pub fn from_slice(slice: &[T]) -> AbiResult<Self>
+        where
+            T: Clone,
+    {
+        if slice.is_empty() {
+            return AbiError::err(AbiErrorKind::Error_empty_type);
+        }
+        Ok(NonEmptyArray(slice.to_vec()))
+    }
```

Comment:
I don't think this function is required.
In the tests you write
```       let state_roots = NonEmptyBytes::from_slice(
            "3a53dc4890241dbe03e486e785761577d1c369548f6b09aa38017828dcdf5c27"
                .decode_hex()
                .unwrap()
                .as_slice(),
        ).unwrap();
```
However, `decode_hex()` returns `Vec<u8>`, and you can just use `NonEmptyArray::new()` instead of this

Reply:
`decode_hex()` returns ` FromHexResult<Data>` already, so it the simplest way to do this.

I have removed the `.as_slice()`. So it looks like this now:
```rust
// bytes32 stateRoots
        let state_roots_hex = "3a53dc4890241dbe03e486e785761577d1c369548f6b09aa38017828dcdf5c27"
            .decode_hex()
            .unwrap()
            .to_vec();
        let state_roots = NonEmptyArray::new(state_roots_hex).unwrap();

        let uint1 = "3402053321874964899321528271743396700217057178612185975187363512030360053932";
        let uint2 = "1235124644010117237054094970590473241953434069965207718920579820322861537001";

        // uint256[2] calldata signatures
        let signatures = Token::Array {
            arr: vec![
                Token::Uint {
                    uint: U256::from_str(uint1).unwrap(),
                    bits: UintBits::new(256).unwrap(),
                },
                Token::Uint {
                    uint: U256::from_str(uint2).unwrap(),
                    bits: UintBits::new(256).unwrap(),
                },
            ],
            kind: ParamType::Uint {
                bits: UintBits::new(256).unwrap(),
            },
        };
        ```

Reply:
`Data` is an alias of `Vec<u8>`, so `.to_vec()` is not needed